processor/groupbyattrsprocessor/attribute_groups.go

```diff
@@ -76,114 +97,150 @@ func newSpansGroupedByAttrs() *spansGroupedByAttrs {
+		if resourceMatches(sgba.At(i).Resource(), referenceAttributes) {
 			return sgba.At(i), true
 		}
 	}
 	return pdata.ResourceSpans{}, false
 }
 
-// resourceMatches verifies if given pdata.Resource matches a composition of another (base) resource and attributes
-func resourceMatches(res pdata.Resource, baseResource pdata.Resource, recordAttrs pdata.AttributeMap) bool {
-	baseAttrs := baseResource.Attributes()
+// findResource searches for an existing pdata.ResourceMetrics that strictly matches with the specified reference
+// Attributes. Returns the matching pdata.ResourceMetrics and bool value which is set to true if found
+func (mgba metricsGroupedByAttrs) findResource(referenceAttributes pdata.AttributeMap) (pdata.ResourceMetrics, bool) {
 
-	// Some attributes in baseResource and recordAttrs might overlap, lets check obvious condition first before iterating
-	minCommonAttrs := baseAttrs.Len() - recordAttrs.Len()
-	if minCommonAttrs < 0 {
-		minCommonAttrs = recordAttrs.Len() - baseAttrs.Len()
-	}
-	maxCommonAttrs := baseAttrs.Len() + recordAttrs.Len()
-	if res.Attributes().Len() > maxCommonAttrs || res.Attributes().Len() < minCommonAttrs {
-		return false
+	for i := 0; i < mgba.Len(); i++ {
+		if resourceMatches(mgba.At(i).Resource(), referenceAttributes) {
+			return mgba.At(i), true
+		}
 	}
+	return pdata.ResourceMetrics{}, false
+}
 
-	matching := true
-	matchedBaseAttrs := 0
-	matchedRecordAttrs := 0
-
-	res.Attributes().Range(func(k1 string, v1 pdata.AttributeValue) bool {
-		if matching {
-			// Prioritize span-level attributes over resource attributes
-			v2, recordAttrFound := recordAttrs.Get(k1)
-			if recordAttrFound {
-				matchedRecordAttrs++
-				if !v1.Equal(v2) {
-					matching = false
-					return true
-				}
-			}
-
-			v2, baseAttrFound := baseAttrs.Get(k1)
-			if baseAttrFound {
-				matchedBaseAttrs++
-				if !v1.Equal(v2) {
-					matching = false
-					return true
-				}
-			}
-
-			if !recordAttrFound && !baseAttrFound {
-				matching = false
-			}
-		}
+// Update the specified (and new) Resource with the properties of the original Resource, and with the
+// required Attributes
+func updateResourceToMatch(newResource pdata.Resource, originResource pdata.Resource, requiredAttributes pdata.AttributeMap) {
+
+	originResource.CopyTo(newResource)
+
+	// This prioritizes required attributes over the original resource attributes, if they overlap
+	attrs := newResource.Attributes()
+	requiredAttributes.Range(func(k string, v pdata.AttributeValue) bool {
+		attrs.Upsert(k, v)
 		return true
 	})
 
-	if matchedBaseAttrs != baseAttrs.Len() || matchedRecordAttrs != recordAttrs.Len() {
-		return false
+}
+
+// findResourceOrElseCreate searches for a Resource with matching attributes and returns it. If nothing is found, it is being created
+func (sgba *spansGroupedByAttrs) findResourceOrElseCreate(originResource pdata.Resource, requiredAttributes pdata.AttributeMap) pdata.ResourceSpans {
```

Comment:
Perhaps `findOrCreateResource` would be a better name?

```suggestion
func (sgba *spansGroupedByAttrs) findOrCreateResource(originResource pdata.Resource, requiredAttributes pdata.AttributeMap) pdata.ResourceSpans {
```